templates/shift-planning-pyomo/main.py

```diff
@@ -0,0 +1,414 @@
+    )
+    parser.add_argument(
+        "-output",
+        default="",
+        help="Path to output file. Default is stdout.",
+    )
+    parser.add_argument(
+        "-duration",
+        default=30,
+        help="Max runtime duration (in seconds). Default is 30.",
+        type=int,
+    )
+    parser.add_argument(
+        "-provider",
+        default="cbc",
+        help="Solver provider. Default is cbc.",
+    )
+    args = parser.parse_args()
+
+    # Read input data, solve the problem, and write the solution.
+    input_data = read_input(args.input)
+    log("Solving shift-planning:")
+    log(f"  - shifts-templates: {len(input_data.get('shifts', []))}")
+    log(f"  - demands: {len(input_data.get('demands', []))}")
+    log(f"  - max duration: {args.duration} seconds")
+    solution = solve(input_data, args.duration, args.provider)
+    write_output(args.output, solution)
+
+
+def solve(
+    input_data: dict[str, Any], duration: int, provider: str
+) -> dict[str, Any]:
+    """Solves the given problem and returns the solution."""
+
+    # Create the Pyomo model
+    model = ConcreteModel()
+
+    # Prepare data
+    shifts, demands = convert_data(input_data)
+    options = input_data.get("options", {})
+
+    # Generate concrete shifts from shift templates.
+    concrete_shifts = get_concrete_shifts(shifts)
+
+    # Determine all unique time periods in which demands occur and the shifts covering them.
+    periods = get_demand_coverage_periods(concrete_shifts, demands)
+
+    # Determine the time we need to cover.
+    required_hours = sum((p.end_time - p.start_time).seconds for p in periods) / 3600
+
+    # Create integer variables indicating how many times a shift is planned.
+    model.x_assign = Var(
+        [(s["id"],) for s in concrete_shifts], within=NonNegativeIntegers
+    )
+
+    # Create variables for tracking various costs.
+    if "under_supply_cost" in options:
+        model.x_under = Var(
+            [(p,) for p in periods], within=NonNegativeIntegers
+        )
+        model.underSupply = Var(within=NonNegativeIntegers)
+    if "over_supply_cost" in options:
+        model.overSupply = Var(within=NonNegativeIntegers)
+    model.shift_cost = Var(within=NonNegativeIntegers)
+
+    # Objective function: minimize the cost of the planned shifts
+    obj_expr = 0
+    if "under_supply_cost" in options:
+        obj_expr += sum(model.x_under[p] for p in periods) * options["under_supply_cost"]
+    if "over_supply_cost" in options:
+        obj_expr += model.overSupply * options["over_supply_cost"]
+    obj_expr += model.shift_cost
+    model.obj = Objective(expr=obj_expr, sense=minimize)
+
+    # Constraints
+
+    # We need to make sure that all demands are covered (or track under supply).
+    for p in periods:
+        constraint_name = f"DemandCover_{p.start_time}_{p.end_time}_{p.qualification}"
+        if constraint_name in model.component_map(Constraint):
+            model.del_component(constraint_name)
```

Comment:
Why is this needed? :thinking:

Reply:
yeah, not sure tbh. There was an error of doubled constraints and a quick search gave this solution. no idea if this is normal or not...